### ISSUE TITLE
linking problem on FreeBSD (12.2).

### ISSUE THREAD

**Issue by sprock** [chapel-lang/chapel#18816]
Hello,

### Summary of Problem

I want to compile (with 1.25.0) some code that was last built using
1.20.  The reference to -resource-dir /usr/lib/clang/10.0.1 looks fishy to me:  that is the system compiler not the llvm11 I installed specifically for chapel. 

chpl -o formula formula_04.chpl --print-commands
<internal clang code generation> -resource-dir /usr/lib/clang/10.0.1 -resource-dir /usr/local/llvm11/lib/clang/11.0.1 -D_GNU_SOURCE=1 -DGASNET_PAR -D_THREAD_SA
FE -I/home/rmason/Software/Chapel/chapel-1.25.0/third-party/gasnet/install/freebsd-amd64-unknown-llvm-none/substrate-udp/seg-everything/include -I/home/rmason/
Software/Chapel/chapel-1.25.0/third-party/gasnet/install/freebsd-amd64-unknown-llvm-none/substrate-udp/seg-everything/include/udp-conduit -DCHPL_JEMALLOC_PREFI
X=chpl_je_ -DCHPL_HAS_GMP -I. -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/localeModels/flat -I/home/rmason/Software/Chapel/chapel-1.25.0/runti
me/include/localeModels -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/comm/gasnet -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/c
omm -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/tasks/fifo -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/threads/pthreads -I/ho
me/rmason/Software/Chapel/chapel-1.25.0/runtime/include/. -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/./qio -I/home/rmason/Software/Chapel/cha
pel-1.25.0/runtime/include/./atomics/cstdlib -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/./mem/jemalloc -I/home/rmason/Software/Chapel/chapel-
1.25.0/third-party/utf8-decoder -I/home/rmason/Software/Chapel/chapel-1.25.0/build/runtime/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/
tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none/include -I/home/rmason/Software/Chapel/chapel
-1.25.0/third-party/jemalloc/install/freebsd-amd64-unknown-llvm-none/include -I/home/rmason/Software/Chapel/chapel-1.25.0/third-party/gmp/install/freebsd-amd64
-unknown-llvm-none/include -I/home/rmason/Software/Chapel/chapel-1.25.0/modules/standard -I/home/rmason/Software/Chapel/chapel-1.25.0/modules/packages -I/home/
rmason/Software/Chapel/chapel-1.25.0/modules/internal -I/tmp//chpl-rmason.deleteme-irCP0m -pthread -DCHPL_GEN_CODE -include sys_basic.h -include ctype.h -inclu
de wctype.h -include llvm/chapel_libc_wrapper.h
# Make Binary - Linking
/usr/local/llvm11/bin/clang++  -resource-dir /usr/lib/clang/10.0.1  /tmp//chpl-rmason.deleteme-irCP0m/chpl__module.o /home/rmason/Software/Chapel/chapel-1.25.0/lib/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none/main.o -o /tmp//chpl-rmason.deleteme-irCP0m/formula.tmp -L/usr/local/lib -L/home/rmason/Software/Chapel/chapel-1.25.0/third-party/jemalloc/install/freebsd-amd64-unknown-llvm-none/lib -Wl,-rpath,/home/rmason/Software/Chapel/chapel-1.25.0/third-party/jemalloc/install/freebsd-amd64-unknown-llvm-none/lib -L/home/rmason/Software/Chapel/chapel-1.25.0/third-party/gmp/install/freebsd-amd64-unknown-llvm-none/lib -Wl,-rpath,/home/rmason/Software/Chapel/chapel-1.25.0/third-party/gmp/install/freebsd-amd64-unknown-llvm-none/lib -L/home/rmason/Software/Chapel/chapel-1.25.0/third-party/re2/install/freebsd-amd64-unknown-llvm-none/lib -Wl,-rpath,/home/rmason/Software/Chapel/chapel-1.25.0/third-party/re2/install/freebsd-amd64-unknown-llvm-none/lib -L/home/rmason/Software/Chapel/chapel-1.25.0/lib/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none -lchpl -lpthread -L/home/rmason/Software/Chapel/chapel-1.25.0/third-party/gasnet/install/freebsd-amd64-unknown-llvm-none/substrate-udp/seg-everything/lib -lgasnet-udp-par -lamudp -pthread -lexecinfo -lm -lm -lgmp -ljemalloc -lm -ldl -lre2 -lpthread -O3 -Wno-unused -Wunused-result -Wno-unused-parameter -Wno-address
ld: error: undefined symbol: re2::RE2::MatchFile(re2::FilePiece&, re2::StringPiece const&, re2::RE2::Anchor, re2::FilePiece*, int) const
>>> referenced by re2-interface.cc
>>>               re2-interface.o:(qio_regex_channel_match) in archive /home/rmason/Software/Chapel/chapel-1.25.0/lib/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none/libchpl.a
clang++: error: linker command failed with exit code 1 (use -v to see
invocation)

Is this a blocking issue with no known work-arounds?  I don't know.  I did not find anything on-line.

### Steps to Reproduce

**Source Code:**
I don't think this has anything to do with the specific sourse code, but here it is.
```chapel
  /* Formula from mass fraction */
  /* chpl -o formula formula_03.chpl */
  /* ./formula -nl <n> --nox=<nanions> --input=<input.in> --control=control.in */
  /* Ouput can be redirected: */
  /* ./formula -nl <n> --nox=<nanions> --input=<input.in> --control=control.in > clinohumite.org */
  /* The table imports nicely to =org= */

  use IO;			// For chapel 1.25.0

  config var nox: int = 1;
  config var input: string;
  config var control: string;

  //  var file: string = 'analyses.txt';
    var shape: 2*int; // for the analyses
    var shaped: 2*int;  // for the data

    var reader = open(input, iomode.r).reader();
    reader.read(shape);
    var analyses: [0..shape[1], 1..shape[1]] real;
    reader.read(analyses);
    reader.close();
    
    reader = open(control, iomode.r).reader();
    reader.read(shaped);
    var oxnames:      [0..shaped[1]] string;
    var ionnames:     [0..shaped[1]] string;
    var molwts:       [0..shaped[1]] real;
    var noxoxide:     [0..shaped[1]] real;
    var nometaloxide: [0..shaped[1]] real;

    reader.read(oxnames);
    reader.read(ionnames);
    reader.read(molwts);
    reader.read(noxoxide);
    reader.read(nometaloxide);
    reader.close();

    var fluorine: int = 0;
    var chlorine: int = 0;
    for i in 0..shaped[1] do {
      if (oxnames(i) == "F") then
	fluorine = i;
      if (oxnames(i) == "Cl") then
	chlorine = i;  
     }

     // Steps 1 & 2
    var s: [0..shape[1]] real;
    var molpropoxide: [0..shape[1], 0..shape[1]] real;
    var atpropoxide: [0..shape[1], 0..shape[1]] real;
    forall (i,j) in {0..shape[1], 0..shape[1]} do {
     molpropoxide(i,j) = analyses(i,j)/molwts(i);
     atpropoxide(i,j) = molpropoxide(i,j) * noxoxide(i);
     s[j] = s[j] + atpropoxide(i,j);
    }

    // Step 3
    var f: [0..shape[1]] real;
    for j in 0..shape[1] do {
      if ( fluorine > 0 ) then
	s[j] = s[j] - atpropoxide(fluorine, j)/2;
      if ( chlorine > 0 ) then
	s[j] = s[j] - atpropoxide(chlorine, j)/2;
      f[j] = 1/s[j];
     }

    // Step 4
    var noanions: [0..shape[1], 0..shape[1]] real;
    forall (i,j) in {0..shape[1], 0..shape[1]} do
     noanions(i,j) = atpropoxide(i,j) * f[j] * nox;  // Command line arg nox

     // Step 5
    var formula: [0..shape[1], 0..shape[1]] real;
    forall (i,j) in {0..shape[1], 0..shape[1]} do
     formula(i,j) = (nometaloxide(i)/noxoxide(i)) * noanions(i,j);

  // Output
  writeln("Formula per ", nox, " anions. \n");
  // Write an org table.
  for i in 0..shaped[1] {
    writef("|%s\t", ionnames(i));
  }
  writef("|\n|");
  for i in 0..shaped[1]-1 {
    writef("---+");  
  }
  writef("---|\n");
  for i in 0..shaped[1] {
    writef("|%7.3dr\t",formula(i,1));	/* Warning: FIXED index! */
  }
  writef("|\n");

```

**Compile command:**
 chpl -o formula formula_04.chpl

### Configuration Information

- Output of `chpl --version`:
- chpl version 1.25.0
  built with LLVM version 11.0.1
Copyright 2020-2021 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
- CHPL_TARGET_PLATFORM: freebsd
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: amd64
CHPL_TARGET_CPU: unknown
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: fifo
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: bundled
CHPL_HWLOC: none
CHPL_RE2: bundled
CHPL_LLVM: system
CHPL_AUX_FILESYS: none
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:

Thanks,
Roger

**Comment by bradcray** [comment]
@sprock: Could you show what the resulting linking problem looks like?  (e.g., the output of the `chpl myProgram.chpl` command)?

**Comment by sprock** [comment]
Hello Brad,
Sorry about that.  I embedded all that in the original submisssion.  Dunno where it went.

chpl -o formula formula_04.chpl --print-commands
<internal clang code generation> -resource-dir /usr/lib/clang/10.0.1 -resource-dir /usr/local/llvm11/lib/clang/11.0.1 -D_GNU_SOURCE=1 -DGASNET_PAR -D_THREAD_SAFE -I/home/rmason/Software/Chapel/chapel-1.25.0/third-party/gasnet/install/freebsd-amd64-unknown-llvm-none/substrate-udp/seg-everything/include -I/home/rmason/Software/Chapel/chapel-1.25.0/third-party/gasnet/install/freebsd-amd64-unknown-llvm-none/substrate-udp/seg-everything/include/udp-conduit -DCHPL_JEMALLOC_PREFIX=chpl_je_ -DCHPL_HAS_GMP -I. -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/localeModels/flat -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/localeModels -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/comm/gasnet -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/comm -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/tasks/fifo -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/threads/pthreads -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/. -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/./qio -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/./atomics/cstdlib -I/home/rmason/Software/Chapel/chapel-1.25.0/runtime/include/./mem/jemalloc -I/home/rmason/Software/Chapel/chapel-1.25.0/third-party/utf8-decoder -I/home/rmason/Software/Chapel/chapel-1.25.0/build/runtime/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none/include -I/home/rmason/Software/Chapel/chapel-1.25.0/third-party/jemalloc/install/freebsd-amd64-unknown-llvm-none/include -I/home/rmason/Software/Chapel/chapel-1.25.0/third-party/gmp/install/freebsd-amd64-unknown-llvm-none/include -I/home/rmason/Software/Chapel/chapel-1.25.0/modules/standard -I/home/rmason/Software/Chapel/chapel-1.25.0/modules/packages -I/home/rmason/Software/Chapel/chapel-1.25.0/modules/internal -I/tmp//chpl-rmason.deleteme-q7ojh8 -pthread -DCHPL_GEN_CODE -include sys_basic.h -include ctype.h -include wctype.h -include llvm/chapel_libc_wrapper.h 

# Make Binary - Linking
/usr/local/llvm11/bin/clang++  -resource-dir /usr/lib/clang/10.0.1  /tmp//chpl-rmason.deleteme-q7ojh8/chpl__module.o /home/rmason/Software/Chapel/chapel-1.25.0/lib/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none/main.o -o /tmp//chpl-rmason.deleteme-q7ojh8/formula.tmp -L/usr/local/lib -L/home/rmason/Software/Chapel/chapel-1.25.0/third-party/jemalloc/install/freebsd-amd64-unknown-llvm-none/lib -Wl,-rpath,/home/rmason/Software/Chapel/chapel-1.25.0/third-party/jemalloc/install/freebsd-amd64-unknown-llvm-none/lib -L/home/rmason/Software/Chapel/chapel-1.25.0/third-party/gmp/install/freebsd-amd64-unknown-llvm-none/lib -Wl,-rpath,/home/rmason/Software/Chapel/chapel-1.25.0/third-party/gmp/install/freebsd-amd64-unknown-llvm-none/lib -L/home/rmason/Software/Chapel/chapel-1.25.0/third-party/re2/install/freebsd-amd64-unknown-llvm-none/lib -Wl,-rpath,/home/rmason/Software/Chapel/chapel-1.25.0/third-party/re2/install/freebsd-amd64-unknown-llvm-none/lib -L/home/rmason/Software/Chapel/chapel-1.25.0/lib/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none -lchpl -lpthread -L/home/rmason/Software/Chapel/chapel-1.25.0/third-party/gasnet/install/freebsd-amd64-unknown-llvm-none/substrate-udp/seg-everything/lib -lgasnet-udp-par -lamudp -pthread -lexecinfo -lm -lm -lgmp -ljemalloc -lm -ldl -lre2 -lpthread -O3 -Wno-unused -Wunused-result -Wno-unused-parameter -Wno-address
ld: error: undefined symbol: re2::RE2::MatchFile(re2::FilePiece&, re2::StringPiece const&, re2::RE2::Anchor, re2::FilePiece*, int) const
>>> referenced by re2-interface.cc
>>>               re2-interface.o:(qio_regex_channel_match) in archive /home/rmason/Software/Chapel/chapel-1.25.0/lib/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none/libchpl.a

This: -resource-dir /usr/lib/clang/10.0.1 looks fishy to me.  I installed llvm11 specifically for chapel.

Thanks,
Roger

**Comment by bradcray** [comment]
Ah, I see, it was in a comment (I hadn't thought to look at the raw sources for the OP).  I edited in hopes of revealing it (but could have missed something else).

**Comment by bradcray** [comment]
@mppf: At a sniff (the mention of `-resource-dir`, this feels like the sort of thing you may have improved between Chapel 1.25.0 and tomorrow's 1.25.1 release, but I'm not confident of that.  Could you take a look at the OP and see what you think?

And/or:

@sprock: If you had the inclination, I'd be curious whether building from our `main` branch would resolve this for you.  As alluded to above, we're doing a 1.25.1 release tomorrow which was largely motivated by various LLVM and clang portability and stability issues that we didn't catch in-house before releasing 1.25.0 (which was the first to use the LLVM back-end by default).

**Comment by mppf** [comment]
I was able to get the point release to work on FreeBSD 12.2 after using `sudo pkg install llvm11` to  installing the llvm11 package. However, I needed the patch below to fix a problem with detecting the installed llvm-config which has a name we had not expected (`llvm-config11` vs `llvm-config-11`) and a bug with setting `CHPL_LLVM_CONFIG` to override it.

I was able to compile `chpl examples/hello.chpl` with this with the default configuration. I don't see a wrong `-resource-dir` with the commands being run. I do see some warnings though like `warning: argument unused during compilation: '--gcc-toolchain=/usr/local'`.

```
diff --git a/util/chplenv/chpl_llvm.py b/util/chplenv/chpl_llvm.py
index 99e918e9..4648ea65 100755
--- a/util/chplenv/chpl_llvm.py
+++ b/util/chplenv/chpl_llvm.py
@@ -110,10 +110,16 @@ def check_llvm_config(llvm_config):
 
 @memoize
 def find_system_llvm_config():
+    llvm_config = overrides.get('CHPL_LLVM_CONFIG', 'none')
+    if llvm_config != 'none':
+        return llvm_config
+
     paths = [ ]
     for vers in llvm_versions():
         paths.append("llvm-config-" + vers + ".0")
         paths.append("llvm-config-" + vers)
+        # this format used by freebsd
+        paths.append("llvm-config" + vers)
         # next ones are for Homebrew
         paths.append("/usr/local/opt/llvm@" + vers + ".0/bin/llvm-config")
         paths.append("/usr/local/opt/llvm@" + vers + "/bin/llvm-config")
```

**Comment by mppf** [comment]
That patch as well as a change to fix the gcc-toolchain unused argument warning is in PR #18820.

**Comment by sprock** [comment]
Hello Brad,

Brad Chamberlain ***@***.***> writes:

> @sprock: If you had the inclination, I'd be curious whether building
> from our main branch would resolve this for you. As alluded to above,
> we're doing a 1.25.1 release tomorrow which was largely motivated by
> various LLVM and clang portability and stability issues that we didn't
> catch in-house before releasing 1.25.0 (which was the first to use the
> LLVM back-end by default).

I just built 7554fdd86a12863b3688d0e5b8f6afdba3be63e3 and applied a
patch to locate llvm_config correctly.  This is the result of compiling
my application:

chpl -o formula formula_04.chpl --print-commands
<internal clang code generation> --gcc-toolchain=/usr/local -D_GNU_SOURCE=1 -DGASNET_PAR -D_THREAD_SAFE -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/gasnet/install/freebsd-amd64-unknown-llvm-none/substrate-udp/seg-everything/include -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/gasnet/install/freebsd-amd64-unknown-llvm-none/substrate-udp/seg-everything/include/udp-conduit -DCHPL_JEMALLOC_PREFIX=chpl_je_ -DCHPL_HAS_GMP -I. -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/runtime/include/localeModels/flat -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/runtime/include/localeModels -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/runtime/include/comm/gasnet -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/runtime/include/comm -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/runtime/include/tasks/fifo -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/runtime/include/threads/pthreads -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/runtime/include/. -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/runtime/include/./qio -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/runtime/include/./atomics/cstdlib -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/runtime/include/./mem/jemalloc -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/utf8-decoder -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/build/runtime/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none/include -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/jemalloc/install/target/freebsd-amd64-unknown-llvm-none/include -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/gmp/install/freebsd-amd64-unknown-llvm-none/include -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/modules/standard -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/modules/packages -I/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/modules/internal -I/tmp//chpl-rmason.deleteme-OPjUSa -pthread -DCHPL_GEN_CODE -include sys_basic.h -include ctype.h -include wctype.h -include llvm/chapel_libc_wrapper.h 
warning: argument unused during compilation: '--gcc-toolchain=/usr/local'

# Make Binary - Linking
/usr/local/llvm11/bin/clang++ /tmp//chpl-rmason.deleteme-OPjUSa/chpl__module.o /home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/lib/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none/main.o -o /tmp//chpl-rmason.deleteme-OPjUSa/formula.tmp --gcc-toolchain=/usr/local -L/usr/local/lib -L/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/jemalloc/install/target/freebsd-amd64-unknown-llvm-none/lib -Wl,-rpath,/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/jemalloc/install/target/freebsd-amd64-unknown-llvm-none/lib -L/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/gmp/install/freebsd-amd64-unknown-llvm-none/lib -Wl,-rpath,/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/gmp/install/freebsd-amd64-unknown-llvm-none/lib -L/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/re2/install/freebsd-amd64-unknown-llvm-none/lib -Wl,-rpath,/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/re2/install/freebsd-amd64-unknown-llvm-none/lib -L/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/lib/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none -lchpl -lpthread -L/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/gasnet/install/freebsd-amd64-unknown-llvm-none/substrate-udp/seg-everything/lib -lgasnet-udp-par -lamudp -pthread -lexecinfo -lm -lm -lgmp -L/home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/third-party/jemalloc/install/target/freebsd-amd64-unknown-llvm-none/lib -ljemalloc -lm -ldl -lre2 -lpthread -O3
clang++: warning: argument unused during compilation: '--gcc-toolchain=/usr/local' [-Wunused-command-line-argument]
ld: error: undefined symbol: re2::RE2::MatchFile(re2::FilePiece&, re2::StringPiece const&, re2::RE2::Anchor, re2::FilePiece*, int) const
>>> referenced by re2-interface.cc
>>>               re2-interface.o:(qio_regex_channel_match) in archive /home/rmason/Software/Chapel/chapel-7554fdd86a12863b3688d0e5b8f6afdba3be63e3/lib/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none/libchpl.a
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
error: Make Binary - Linking

I'll try gcc instead.

Cheers,
Roger

**Comment by mppf** [comment]
@sprock -  I haven't been able to reproduce this link error. Could you try rebuilding RE2 with
```
cd third-party/re2
make clobber
make
```

and check to see if the `clang++` compiler used here is different from the one from your `chpl` invocation?

**Comment by sprock** [comment]
Michael Ferguson ***@***.***> writes:

> @sprock - I haven't been able to reproduce this link error. Could you try rebuilding RE2 with
>
> cd third-party/re2
> make clobber
> make
>
> and check to see if the clang++ compiler used here is different from the one from your chpl invocation?

It is using /usr/local/llvm11/bin/clang++

**Comment by mppf** [comment]
Also I suspect you aren't actually using regular expressions. Does your program compile if you use `export CHPL_RE2=none` and then run `make` again?

**Comment by sprock** [comment]
Hello Michael,
Michael Ferguson ***@***.***> writes:

> Also I suspect you aren't actually using regular expressions. Does your program compile if you use export CHPL_RE2=none and then run make again?

Yes, it does.

I have done some furter testing.

source chapel-7554f/util/setchplenv.bash

export CHPL_RE2=bundled
export CHPL_COMM=gasnet

machine info: FreeBSD pyrope 12.2-RELEASE FreeBSD 12.2-RELEASE r366954 GENERIC amd64
CHPL_HOME: /usr/home/rmason/Software/Chapel/chapel-7554f
script location: /usr/home/rmason/Software/Chapel/chapel-7554f/util/chplenv
CHPL_TARGET_PLATFORM: freebsd
CHPL_TARGET_COMPILER: clang
CHPL_TARGET_ARCH: amd64
CHPL_TARGET_CPU: unknown
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: fifo
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: none
CHPL_HWLOC: none
CHPL_RE2: bundled *
CHPL_LLVM: unset
CHPL_AUX_FILESYS: none

**** Build
1.  patch < freebsd-patch-1.patch (to find system llvm11).
2.  gmake -j2
3.  ln -s /home/rmason/Software/Chapel/chapel-7554f/bin/freebsd-amd64/chpl ~/bin/chpl

**** Test hello
1.  chpl -o hello examples/hello6-taskpar-dist.chpl
2.  export GASNET_SPAWNFN=S
3.  export GASNET_SSH_SERVERS="pyrope forsterite" (n.b. forsterite has
    FreeBSD 13.0 c.f. 12.2).
4.  ./hello -nl 2
ld-elf.so.1: Shared object "libre2.so.9" not found, required by "hello_real"
zsh: terminated  ./hello -nl 2
That is OK, it is probably from =forsterite=: =libre2= is not installed there.
5.  =ldd= shows that =hello_real= is linked against
    =/usr/local/lib/libre2.so.9=, contradicting the envar setup.
6.  export GASNET_SSH_SERVERS="pyrope braidperthite"
7.  ./hello -nl 2                                   
Hello, world! (from locale 0 of 2 named pyrope)
Hello, world! (from locale 1 of 2 named braidperthite)
Good, that works, but why is it linked against =libre2= at all and why
is it linked against the wrong version of that library?

**** Test formula
1.  chpl -o formula formula_04.chpl

clang++: warning: argument unused during compilation: '--gcc-toolchain=/usr/local' [-Wunused-command-line-argument]
ld: error: undefined symbol: re2::RE2::MatchFile(re2::FilePiece&, re2::StringPiece const&, re2::RE2::Anchor, re2::FilePiece*, int) const
>>> referenced by re2-interface.cc
>>>               re2-interface.o:(qio_regex_channel_match) in archive /usr/home/rmason/Software/Chapel/chapel-7554f/lib/freebsd/llvm/amd64/arch-unknown/loc-flat/comm-gasnet/udp/everything/tasks-fifo/tmr-generic/unwind-none/mem-jemalloc/atomics-cstdlib/none/bundled/fs-none/lib_pic-none/sanitizers-none/libchpl.a
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
error: Make Binary - Linking

I don't understand.  =formula_04.chpl= does not use =libre2=.

Cheers,
Roger

**Comment by mppf** [comment]
Your printchplenv is showing `CHPL_RE2: bundled *` -- that means that the Chapel *runtime* is configured to work with the bundled RE2. It doesn't matter if your program uses RE2 for this part. To turn it off you need to do `export CHPL_RE2=none` or equivalent setenv.

I think the problem here is that you have a system-wide installation of RE2 and the build process is finding that but not finding the one built in third-party/re2. Which is a bug on our part. So far I haven't been able to reproduce it, though. However I do see the `ldd` showing the system libre2.so and at least that's something I can investigate.

**Comment by bradcray** [comment]
Hi @sprock — I lost track of this issue a bit, but wanted to check in and see whether this was holding you up in any way, and whether Michael's suggestion to just turn off re2 for now (assuming you aren't using it) works.  GitHub doesn't notify about mentions, but I wanted to point out that Michael has been wrestling with the more general issue that is causing this behavior on #18840 .  There was also a PR that was merged related to this (#18820 above), though it's not obvious to me that this was sufficient to resolve this error (I'd guess not).

**Comment by sprock** [comment]
Hello Brad,

Brad Chamberlain ***@***.***> writes:

> Hi @sprock — I lost track of this issue a bit, but wanted to check in
> and see whether this was holding you up in any way, and whether
> Michael's suggestion to just turn off re2 for now (assuming you aren't
> using it) works. GitHub doesn't notify about mentions, but I wanted to
> point out that Michael has been wrestling with the more general issue
> that is causing this behavior on #18840 . There was also a PR that was
> merged related to this (#18820 above), though it's not obvious to me
> that this was sufficient to resolve this error (I'd guess not).

Sorry for the lack of response - I had to stop and work on something
else because of a deadline.

I can now confirm that turning off re2 during the build of chapel allows my
application to compile.

Thanks for the help.

**Comment by sprock** [comment]
Roger Mason ***@***.***> writes:

> I can now confirm that turning off re2 during the build of chapel allows my
> application to compile.

Unfortunately I'm now getting a runtime error:

uncaught BadFormatError: bad format: not a floating point number (while reading real(64) with path "unknown" offset 57)
  formula_05.chpl:34: thrown here
  formula_05.chpl:34: uncaught here

The source code is in my original post.  Line 43 is:

reader.read(molwts);

Thanks again.

**Comment by bradcray** [comment]
Hi @sprock —

That error suggests to me that the attempt to read a `real` in that line failed because it didn't fine something that looked like a `real` in the input stream.  We'd need your input file/stream in order to determine whether that's reasonable or not.  I might suggest printing out the things you read just prior to this line to make sure that the rest of the input has been processed as you'd expect; or checking whether that offset 57 points to the place in your file that you'd expect at this point in the input.

**Comment by mppf** [comment]
I think this is resolved temporarily in #18820 and #18880 will have the better, longer term fix. I'll mark #18880 to close this when it merges.
